server/src/main/java/io/deephaven/server/table/ops/JoinTablesGrpcImpl.java

```diff
@@ -86,8 +84,11 @@ public Table create(final T request, final List<SessionState.ExportObject<Table>
         if (!lhs.isRefreshing() && !rhs.isRefreshing()) {
             result = realTableOperation.apply(lhs, rhs, columnsToMatch, columnsToAdd, request);
         } else {
-            result = updateGraphProcessor.sharedLock().computeLocked(
-                    () -> realTableOperation.apply(lhs, rhs, columnsToMatch, columnsToAdd, request));
+            final AwareFunctionalLock sharedLock = lhs.isRefreshing()
+                    ? lhs.getUpdateContext().getSharedLock()
+                    : rhs.getUpdateContext().getExclusiveLock();
```

Comment:
This should always be the shared lock. And you want to be sure both parents have the same UGP. I think you want to use `getUpdateContext(lhs, rhs)` earlier to identify the lock to use.